File: python3/002. Add Two Numbers/AddTwoNumbers.py
# 2. Add Two Numbers
# Accepted 96ms

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        ret = ListNode(-1)
        retHead = ret
        carry = 0
        
        while l1 != None or l2 != None or carry != 0:
            a = 0 if l1 == None else l1.val
            b = 0 if l2 == None else l2.val
            
            sum = a + b + carry
            carry = sum // 10
            
            ret.next = ListNode(sum % 10)
            ret = ret.next
            
            l1 = None if l1 == None else l1.next
            l2 = None if l2 == None else l2.next
        
        return retHead.next

File: src/javascript/002. Add Two Numbers/addTwoNumbers.js
// 2. Add Two Numbers
// Accepted 124ms

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let ret = new ListNode(-1);
    let retHead = ret;
    let carry = 0;
    
    while(l1 != null || l2 != null || carry != 0) {
        let a = (l1 == null) ? 0 : l1.val;
        let b = (l2 == null) ? 0 : l2.val;
        
        let newNode = new ListNode((a + b + carry) % 10);
        carry = parseInt(Math.floor((a + b + carry) / 10));
        
        ret.next = newNode;
        ret = ret.next;
        
        if(l1 != null)
            l1 = l1.next;
        if(l2 != null)
            l2 = l2.next;
    }
    
    return retHead.next;
};
